### PR TITLE
Fix expose_stan_functions under Stanc 2.31

### PR DESCRIPTION
**Pull request by andrjohns**
Part of the `expose_stan_functions_hacks` code was replacing `= nullptr` with `= 0`, intended for the `pstream__ = nullptr` default argument. However this causes a compilation failure under 2.31, since the generated code now includes default template parameters with `nullptr` which are then replaced.

This PR updates the regex to only replace the `nullptr` if it's followed by a parenthesis - as the function definitions will always generate the `pstream__ = nullptr` as the last argument. I've also added a basic test.

This will also need to be backported to 2.26

### CODE REVIEW

**Comment by andrjohns** [comment]
I've also verified that this fixes the test failure for the `prophet` package under 2.31

File: rstan/rstan/R/expose_stan_functions.R [hunk 1]
```diff
@@ -31,7 +31,7 @@ expose_stan_functions_hacks <- function(code, includes = NULL) {
   code <- gsub("stan::math::accumulator<double>& lp_accum__, std::ostream* pstream__ = nullptr){",
                "std::ostream* pstream__ = nullptr){\nstan::math::accumulator<double> lp_accum__;",
                code, fixed = TRUE)
-  code <- gsub("= nullptr", "= 0", code, fixed = TRUE)
+  code <- gsub("= nullptr)", "= 0)", code, fixed = TRUE)
```
Comment (hsbadr) [inline rstan/rstan/R/expose_stan_functions.R]: Is there other patterns that we need to support here, such as `"nullptr,"`, `",nullptr"`, or `", nullptr"`?

Comment (andrjohns) [reply rstan/rstan/R/expose_stan_functions.R]: I would have said no, but the test failure under ubuntu is making me question that. I'll investigate and update the PR

**Comment by hsbadr** [comment]
Does this fix https://github.com/stan-dev/rstan/issues/954?

File: rstan/rstan/R/expose_stan_functions.R (hunk 1 above)
Comment (hsbadr) [reply rstan/rstan/R/expose_stan_functions.R:34]: Can we just exclude `pstream__ = nullptr`?
```suggestion
  code <- gsub("pstream__ = nullptr", "pstream__ =nullptr", code, fixed = TRUE)
  code <- gsub("= nullptr", "= 0", code, fixed = TRUE)
  code <- gsub("pstream__ =nullptr", "pstream__ = nullptr", code, fixed = TRUE)
```

**Comment by andrjohns** [comment]
It's bizarre, I can't replicate the test failures locally under MacOS, Ubuntu, or Windows

**Comment by andrjohns** [comment]
Narrowed it down a bit more. The test will fail when run through `rcmdcheck()` but not when run through `test_file()`

**Comment by andrjohns** [comment]
Ah, it looks like `sourceCpp`-based functions aren't compatible with `testthat`, since it modifies include paths: https://stat.ethz.ch/pipermail/r-package-devel/2017q3/001951.html

May have to just remove the test entirely (after testing locally)

**Comment by hsbadr** [comment]
> it looks like `sourceCpp`-based functions aren't compatible with `testthat`,

I see. Could probably use Catch: https://testthat.r-lib.org/reference/use_catch.html.

Anyway, it works on my local machine.

**Review by hsbadr — APPROVED** [review]
LGTM

**Pull request merged**